hw/xbox/nv2a.c

```diff
@@ -5012,36 +5015,89 @@ static void pgraph_method(NV2AState *d,
 
     case NV097_SET_VERTEX_DATA2F_M ...
             NV097_SET_VERTEX_DATA2F_M + 0x7c: {
-        slot = (class_method - NV097_SET_VERTEX_DATA2F_M) / 8;
-        unsigned int part =
-            ((class_method - NV097_SET_VERTEX_DATA2F_M) % 8) / 4;
-        pg->vertex_attributes[slot].inline_value[part] = *(float*)&parameter;
-        pg->vertex_attributes[slot].inline_value[2] = 0.0;
-        pg->vertex_attributes[slot].inline_value[3] = 1.0;
+        slot = (class_method - NV097_SET_VERTEX_DATA2F_M) / 4;
+        unsigned int part = slot % 2;
+        slot /= 2;
+        VertexAttribute *attribute = &pg->vertex_attributes[slot];
+        allocate_inline_buffer_vertices(pg, slot);
+        attribute->inline_value[part] = *(float*)&parameter;
+        /* FIXME: Should these really be set to 0.0 and 1.0 ? Conditions? */
+        attribute->inline_value[2] = 0.0;
+        attribute->inline_value[3] = 1.0;
+        if ((slot == 0) && (part == 1)) {
+            finish_inline_buffer_vertex(pg);
+        }
         break;
     }
-
     case NV097_SET_VERTEX_DATA4F_M ...
             NV097_SET_VERTEX_DATA4F_M + 0xfc: {
-        slot = (class_method - NV097_SET_VERTEX_DATA4F_M) / 16;
-        unsigned int part =
-            ((class_method - NV097_SET_VERTEX_DATA4F_M) % 16) / 4;
-        pg->vertex_attributes[slot].inline_value[part] = *(float*)&parameter;
+        slot = (class_method - NV097_SET_VERTEX_DATA4F_M) / 4;
+        unsigned int part = slot % 4;
+        slot /= 4;
```

Comment:
Reduces the risk of specifying a wrong value in the subtraction, the slot calculation is already prone to that.
Personally, I like this better. A more common slot calculation macro would probably be best.